src/libraries/System.IO.Packaging/src/System/IO/Packaging/InterleavedZipPackagePartStream.cs

```diff
@@ -0,0 +1,572 @@
+                return _currentOffset;
+            }
+            set
+            {
+                CheckClosed();
+                Seek(value, SeekOrigin.Begin);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override long Length
+        {
+            get
+            {
+                CheckClosed();
+                Debug.Assert(CanSeek);
+
+                long length = 0;
+                for (int pieceNumber = 0; pieceNumber < _dir.GetNumberOfPieces(); ++pieceNumber)
+                {
+                    checked { length += _dir.GetStream(pieceNumber).Length; }
+                }
+                return length;
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override void Dispose(bool disposing)
+        {
+
+
+            try
+            {
+                if (disposing)
+                {
+                    if (!_closed)
+                    {
+                        _dir.Close();
+                    }
+                }
+            }
+            finally
+            {
+                _closed = true;
+                base.Dispose(disposing);
+            }
+        }
+
+        private void CheckClosed()
+        {
+            if (_closed)
+                throw new ObjectDisposedException(null, SR.StreamObjectDisposed);
+        }
+
+        /// <summary>
+        /// Infer the current piece number from _currentOffset.
+        /// </summary>
+        /// <remarks>
+        /// Storing the current piece number in a field and computing the current offset from it
+        /// would also have been possible, but less efficient.
+        /// </remarks>
+        private int GetCurrentPieceNumber()
+        {
+            // Since this property is likely to be read more often than _currentOffset
+            // gets updated, its value is cached in _currentPieceNumber.
+            // The validity of the cached value is monitored using _offsetForCurrentPieceNumber.
+            if (_offsetForCurrentPieceNumber != _currentOffset)
+            {
+                // Cached value is stale. Refresh.
+                _currentPieceNumber = _dir.GetPieceNumberFromOffset(_currentOffset);
+                _offsetForCurrentPieceNumber = _currentOffset;
+            }
+            return _currentPieceNumber;
+        }
+
+        /// <summary>
+        /// Moves a piece stream at position zero to its new absolute position by reading from it.
+        /// </summary>
+        private static void SeekUnderlyingPieceStream(Stream pieceStream, long byteCount)
+        {
+            const int BufferSize = 4096;
+            long remainingBytes = byteCount;
+            byte[] readBuffer = ArrayPool<byte>.Shared.Rent(BufferSize);
+            int bytesRead = byteCount < BufferSize ? (int)byteCount : BufferSize;
+
+            try
+            {
+                do
+                {
+                    bytesRead = pieceStream.Read(readBuffer, 0, bytesRead);
```

Comment:
The buffer length calculation looks incorrect here. It reads at most BufferSize.
Pseudo code should be:

```
remainingBytes = arg1;
byte[] buffer = Rent(BufferSize);
do
{
    // note that buffer can be larger than requested size and we can utilize it
    bytesRead = pieceStream.Read(buffer, 0, Min(buffer.Length, remainingBytes);
    remainingBytes -= bytesRead;
} while (remainingBytes > 0 && bytesRead > 0);
```